doc/neps/nep-0052-python-api-cleanup.rst

```diff
@@ -0,0 +1,333 @@
+be hidden by not including it in ``__dir__``. In case of hiding, a ``..
+legacy::`` directory may be used to mark such functionality in the
+documentation.
+
+A test will be added to ensure limited future growth of all namespaces; i.e.,
+every new entry will need to be explicitly added to an allow-list.
+
+
+Cleaning up the submodule structure
+```````````````````````````````````
+
+We will clean up the NumPy submodule structure, so it is easier to navigate.
+When this was discussed before (see
+`MAINT: Hide internals of np.lib to only show submodules <https://github.com/numpy/numpy/pull/18447>`__)
+there was already rough consensus on that - however it was hard to pull off in
+a minor release.
+
+A basic principle we will adhere to is "one function, one location". Functions
+that are exposed in more than one namespace (e.g., many functions are present
+in ``numpy`` and ``numpy.lib``) need to find a single home.
+
+We will reorganize the API reference guide along main and submodule namespaces,
+and only within the main namespace use the current subdivision along
+functionality groupings. Also by "mainstream" and special-purpose namespaces:
+
+::
+
+    # Regular/recommended user-facing namespaces for general use. Present these
+    # as the primary set of namespaces to the users.
+    numpy
+    numpy.exceptions
+    numpy.fft
+    numpy.linalg
+    numpy.polynomial
+    numpy.random
+    numpy.testing
+    numpy.typing
+
+    # Special-purpose namespaces. Keep these, but document them in a separate
+    # grouping in the reference guide and explain their purpose.
+    numpy.array_api
+    numpy.ctypeslib
+    numpy.emath
+    numpy.f2py  # only a couple of public functions, like `compile` and `get_include`
+    numpy.lib.stride_tricks
+    numpy.rec
+    numpy.types
+
+    # Legacy (prefer not to use, there are better alternatives and/or this code
+    # is deprecated or isn't reliable). This will be a third grouping in the
+    # reference guide; it's still there, but de-emphasized and the problems
+    # with it or better alternatives are explained in the docs.
+    numpy.char
+    numpy.distutils
+    numpy.ma
+
+    # To remove
+    numpy.compat
+    numpy.core  # rename to _core
+    numpy.doc
+    numpy.math
+    numpy.matlib
+    numpy.matrixlib
+    numpy.version
+
+    # To clean out or somehow deal with: everything in `numpy.lib`
+
+.. note::
+
+    TBD: will we preserve ``np.lib`` or not? It only has a couple of unique
+    functions/objects, like ``Arrayterator`` (a candidate for removal) and the
+    ``stride_tricks`` subsubmodule. ``numpy.lib`` itself is not a coherent
+    namespace, and does not even have a reference guide page.
+
+We will make all submodules available lazily, so that users don't have to type
+``import numpy.xxx`` but can use ``import numpy as np; np.xxx.*``, while at the
+same time not negatively impacting the overhead of ``import numpy``. This has
+been very helpful for teaching scikit-image and SciPy.
```

Comment:
This also seems friendly towards Spyder victims. At least I've seen situations a while back where code using e.g. `scipy.optimize`  (or similar) worked without the `scipy.optimize` submodule import, but only in Spyder where it was developed. The same script when run in vanilla Python raised. This is a clear Spyder issue but still this sounds very helpful.

Reply:
Thanks @adeak, I'll add that to the motivation.

Reply:
To be fair my information comes from 2018. Spyder may have changed strategy since.

Reply:
I'm happy and sad to say that I'm not libelling Spyder. I tried with Spyder 5.1.5 and `import logging` will import `logging.handlers`. And if there's a setting to change this default I can't find it by perusing the settings. So my remark stil applies.

Reply:
Thanks, I've included this now.